library/std/src/sys/pal/unix/net.rs

```diff
@@ -47,7 +46,9 @@ pub fn cvt_gai(err: c_int) -> io::Result<()> {
 
     #[cfg(not(target_os = "espidf"))]
     let detail = unsafe {
-        str::from_utf8(CStr::from_ptr(libc::gai_strerror(err)).to_bytes()).unwrap().to_owned()
+        // We can't always expect a UTF-8 environment. When we don't get that luxury,
+        // it's better to give a low-quality error message than none at all.
+        String::from_utf8_lossy(CStr::from_ptr(libc::gai_strerror(err)).to_bytes())
```

Comment:
Perhaps use [`CStr::to_string_lossy`](https://doc.rust-lang.org/std/ffi/struct.CStr.html#method.to_string_lossy) here instead, which will avoid allocating a new `String` if not necessary?

Reply:
[`String::from_ut8_lossy`](https://doc.rust-lang.org/nightly/std/string/struct.String.html#method.from_utf8_lossy) also avoids allocating a new  `String` if necessary; it returns a `Cow<'_, str>` to handle that. It looks like [`CStr::to_string_lossy`](https://doc.rust-lang.org/src/alloc/ffi/c_str.rs.html#1100) just delegates to `String::from_ut8_lossy` anyway.

Reply:
Oh, quite right I hadn't realised that!  Ignore me 😳